### PR TITLE
fix: make challenge links on profile page work

### LINKED ISSUE #36085: Broken links on profile page
**Describe your problem and how to reproduce it:**
the challenge links in the profile timeline are broken.

image

### PR DESCRIPTION
**Pull request by moT01**
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #36085

### CODE REVIEW

**Comment by ojeytonwilliams** [comment]
I'll take a proper look at this soon, but my first impressions are that this is a nice solution.  The extra work the server has to do is pretty small.

I am still wondering if we can get away with not using the server at all.  Nonetheless, if we are using it, this looks like the way to go.

**Comment by raisedadead** [comment]
We can update the styles. It was sort of a mess. Feel free to take a dig at it.

**Comment by ojeytonwilliams** [comment]
@raisedadead The issue is that the TimeLine's cert links are currently 404s, so we were wondering if they should be filtered out or made to point at the certifications.

Also, @RandellDawson noticed that the certifications appear in the TimeLine even if they aren't public.  Presumably that's a bug!

**Comment by raisedadead** [comment]
🤦‍♂ I need more coffee. Yes, this looks like a bug.

**Comment by ahmadabdolsaheb** [comment]
it seems like we are moving in the direction of completely removing these links and making the the profile more resume like. I wonder if we should sunset this list now if fixing it is too time consuming. @raisedadead thoughts?

**Comment by raisedadead** [comment]
> it seems like we are moving in the direction of completely removing these links and making the the profile more resume like. I wonder if we should sunset this list now if fixing it is too time consuming. 

I think this should be fixed now. 

We have not planned or designed anything about the "resume" like settings page. It could presumably take weeks. From my point of view this is a blocker issue right now from releasing master.

I would defer to @QuincyLarson though.

File: client/src/components/profile/components/TimeLine.js [hunk 1]
```diff
@@ -42,7 +41,7 @@ const propTypes = {
     })
   ),
   fetchIdToNameMap: PropTypes.func.isRequired,
-  idToNameMap: PropTypes.objectOf(PropTypes.string),
+  idToNameMap: PropTypes.object,
```
Comment (ojeytonwilliams) [inline client/src/components/profile/components/TimeLine.js]:
```suggestion
  idToNameMap: PropTypes.objectOf(
    PropTypes.shape({
      challengePath: PropTypes.string,
      challengeTitle: PropTypes.string
    })
  ),
```
Might as well specify the type of object.

File: client/src/components/profile/components/TimeLine.js [hunk 2]
```diff
@@ -70,13 +69,12 @@ class Timeline extends Component {
   renderCompletion(completed) {
     const { idToNameMap } = this.props;
     const { id, completedDate } = completed;
-    const challengeDashedName = idToNameMap[id];
+    const challengeTitle = idToNameMap[id].challengeTitle;
```
Comment (ojeytonwilliams) [inline client/src/components/profile/components/TimeLine.js]:
```suggestion
    const { challengeTitle, challengePath } = idToNameMap[id];
```
Just to save a line.

File: client/src/components/profile/components/TimeLine.js [hunk 3]
```diff
@@ -70,13 +69,12 @@ class Timeline extends Component {
...
+    const challengePath = idToNameMap[id].challengePath;
```
Comment (ojeytonwilliams) [inline client/src/components/profile/components/TimeLine.js]:
```suggestion
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
I have a couple of nitpicks, but otherwise this looks good.

**Review by ojeytonwilliams — APPROVED** [review]

**Review by raisedadead — APPROVED** [review]

**Comment by moT01** [comment]
Thanks for all the reviews lately everyone! 🎉 @raisedadead @ahmadabdolsaheb @RandellDawson @ojeytonwilliams @thecodingaviator @Manish-Giri @jonathan-grah 🎉

**Pull request merged**